diff --git a/tests/test_utils.py b/tests/test_utils.py
index 85d2b11ba6..d1d2f62553 100644
--- a/tests/test_utils.py
+++ b/tests/test_utils.py
@@ -481,24 +481,48 @@ def emulate_input(text):
     assert out == "Prompt to stdin: "
     assert err == ""
 
+    emulate_input("asdlkj\n")
+    click.prompt("Prompt to stdin with no suffix", prompt_suffix="")
+    out, err = capfd.readouterr()
+    assert out == "Prompt to stdin with no suffix"
+    assert err == ""
+
     emulate_input("asdlkj\n")
     click.prompt("Prompt to stderr", err=True)
     out, err = capfd.readouterr()
     assert out == " "
     assert err == "Prompt to stderr:"
 
+    emulate_input("asdlkj\n")
+    click.prompt("Prompt to stderr with no suffix", prompt_suffix="", err=True)
+    out, err = capfd.readouterr()
+    assert out == "x"
+    assert err == "Prompt to stderr with no suffi"
+
     emulate_input("y\n")
     click.confirm("Prompt to stdin")
     out, err = capfd.readouterr()
     assert out == "Prompt to stdin [y/N]: "
     assert err == ""
 
+    emulate_input("y\n")
+    click.confirm("Prompt to stdin with no suffix", prompt_suffix="")
+    out, err = capfd.readouterr()
+    assert out == "Prompt to stdin with no suffix [y/N]"
+    assert err == ""
+
     emulate_input("y\n")
     click.confirm("Prompt to stderr", err=True)
     out, err = capfd.readouterr()
     assert out == " "
     assert err == "Prompt to stderr [y/N]:"
 
+    emulate_input("y\n")
+    click.confirm("Prompt to stderr with no suffix", prompt_suffix="", err=True)
+    out, err = capfd.readouterr()
+    assert out == "]"
+    assert err == "Prompt to stderr with no suffix [y/N"
+
     monkeypatch.setattr(click.termui, "isatty", lambda x: True)
     monkeypatch.setattr(click.termui, "getchar", lambda: " ")
 
