diff --git a/tests/integration/test_projects_using_isort.py b/tests/integration/test_projects_using_isort.py
index 515fc0841..61f66e7e7 100644
--- a/tests/integration/test_projects_using_isort.py
+++ b/tests/integration/test_projects_using_isort.py
@@ -13,6 +13,8 @@
 from subprocess import check_call
 from typing import Generator, Sequence
 
+import pytest
+
 from isort.main import main
 
 
@@ -26,6 +28,10 @@ def run_isort(arguments: Generator[str, None, None] | Sequence[str]):
     main(["--check-only", "--diff", *arguments])
 
 
+@pytest.mark.xfail(
+    reason="Project is incorrectly formatted after PR #2236, should be fixed "
+    "after a release and the project formatting again."
+)
 def test_django(tmpdir):
     git_clone("https://github.com/django/django.git", tmpdir)
     run_isort(
diff --git a/tests/unit/profiles/test_black.py b/tests/unit/profiles/test_black.py
index 8f7501005..209d83d82 100644
--- a/tests/unit/profiles/test_black.py
+++ b/tests/unit/profiles/test_black.py
@@ -447,8 +447,8 @@ def sub(a: np.ndarray, b: np.ndarray) -> np.ndarray: ...
 
 def test_black_trailing_comma():
     black_test(
-    "from x import (a, b, c,)\n",
-    """\
+        "from x import (a, b, c,)\n",
+        """\
 from x import (
     a,
     b,
diff --git a/tests/unit/test_isort.py b/tests/unit/test_isort.py
index 29834bd32..8a256fbad 100644
--- a/tests/unit/test_isort.py
+++ b/tests/unit/test_isort.py
@@ -5576,6 +5576,18 @@ def test_split_on_trailing_comma() -> None:
     assert output == expected_output
 
 
+def test_split_on_trailing_comma_wih_as() -> None:
+    test_input = "from lib import (a as b,)"
+    expected_output = """from lib import a as b
+"""
+
+    output = isort.code(test_input, split_on_trailing_comma=True)
+    assert output == expected_output
+
+    output = isort.code(expected_output, split_on_trailing_comma=True)
+    assert output == expected_output
+
+
 def test_infinite_loop_in_unmatched_parenthesis() -> None:
     test_input = "from os import ("
 
