diff --git a/tests/test_more.py b/tests/test_more.py
index 6b78ca53..4673109d 100644
--- a/tests/test_more.py
+++ b/tests/test_more.py
@@ -1448,6 +1448,10 @@ def test_max_split(self):
                 ('a,b,c,d', lambda c: c != ',', 2),
                 [['a'], [',', 'b'], [',', 'c', ',', 'd']],
             ),
+            (
+                ([1], lambda x: x == 1, 1),
+                [[1]],
+            ),
         ]:
             actual = list(mi.split_after(*args))
             self.assertEqual(actual, expected)
